lib/dockerregistry/checks.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inventory
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strings"
+
+	gogit "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// ImageRemovalCheck implements the PreCheck interface and checks against
+// pull requests that attempt to remove any images from the promoter manifests.
+type ImageRemovalCheck struct {
+	GitRepoPath    string
+	MasterSHA      plumbing.Hash
+	PullRequestSHA plumbing.Hash
+}
+
+func isValidSha(potenitalSHA string) error {
+	sha256Length := 64
+	if len(potenitalSHA) == sha256Length {
+		_, err := hex.DecodeString(potenitalSHA)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	return fmt.Errorf("Length of SHA is %d characters, should be 64",
+		len(potenitalSHA))
+}
+
+// MKRealImageRemovalCheck returns an instance of ImageRemovalCheck.
+func MKRealImageRemovalCheck(
+	gitRepoPath string,
+) (*ImageRemovalCheck, error) {
+	masterSHAStr := os.Getenv("PULL_BASE_SHA")
+	err := isValidSha(masterSHAStr)
+	if err != nil {
+		return nil, fmt.Errorf("The PULL_BASE_SHA environment variable "+
+			"is invalid: %v", err)
+	}
+	pullRequestSHAStr := os.Getenv("PULL_PULL_SHA")
+	err = isValidSha(pullRequestSHAStr)
+	if err != nil {
+		return nil, fmt.Errorf("The PULL_PULL_SHA environment variable "+
+			"is invalid: %v", err)
+	}
+	masterSHA := plumbing.NewHash(masterSHAStr)
+	pullRequestSHA := plumbing.NewHash(pullRequestSHAStr)
+	return &ImageRemovalCheck{
+		gitRepoPath,
+		masterSHA,
+		pullRequestSHA,
+	}, nil
+}
+
+// Run executes ImageRemovalCheck on a set of promotion edges.
+// Returns an error if the pull request removes images from the
+// promoter manifests.
+func (check *ImageRemovalCheck) Run(
+	edges map[PromotionEdge]interface{},
+) error {
+	var err error
+	// Open the Git repo of the current working directory, check
+	// out the master branch of the repo, parse and generate promotion
+	// edges from the manifests, and then reset the working directory
+	// by checking out the pull request branch of the Git repo. This
+	// is done to make sure that this check is not leaving any lasting
+	// effects on the working directory when the check finishes.
+	r, err := gogit.PlainOpen(check.GitRepoPath)
+	if err != nil {
+		return fmt.Errorf("Could not open the Git repo: %v", err)
+	}
+
+	// Check out the master branch of the repo
```

Comment:
This comment is a little too low-level to be useful. Instead try to think one level higher (explain what you're doing here to someone unfamiliar with the codebase). When in doubt, try to explain *why* something is done; if that's difficult to express, then next best thing is to explain a high-level overview of why some piece of code *exists* at all. What's its *purpose*? Another way to think about it is --- if you were to move this section of code to its own function, what would be the best name for the function?

Reply:
This is a really good trick!